### PR TITLE
feat!: Upgrade rev.2

### PR DESCRIPTION
**Pull request by digimaun**
* Checkpoint PR to merge upgrade revision 2.
  * Replaces existing implementation. Module temporarily called upgrade2.
* Upgrade allows configuration of settings, version and release train for any AIO related extension.
* Unit tests adding 98% code coverage, with Test Scenario pattern established in prior work module unit test refresh.
* Adds `parse_kvp_nargs` util function. This should be the go forward for parsing any kvp nargs based args we support. It is superior to the carry over `assemble_nargs_to_dict` in every way. `assemble_nargs_to_dict` should eventually be deleted.
* Table display "The Upgrade Story" by default which shows a clear upgrade strategy.

Follow ups:
- System extension TBD.
- Some handling of non-success starting state of an extension.
- Look into extension status and determine if a status of error can occur with a success http response code. 
- Future of packaging use.
- QOL and nits.

### CODE REVIEW

File: azext_edge/edge/_help.py [hunk 1]
```diff
@@ -574,24 +574,27 @@ def load_iotops_help():
...
-        - name: Upgrade the instance and specify the schema registry resource Id.
+        - name: Set extension config settings that apply should be during upgrade.
+           To remove a setting provide the key with no value.
+          text: >
+            az iot ops upgrade --name myinstance -g myresourcegroup --ops-config key1=value1 deletekey
+        - name: Provide an explicit IoT Operations version or release train to upgrade to.
+            Not recommended for typical use cases.
           text: >
-            az iot ops upgrade --name myinstance -g myresourcegroup --sr-resource-id $SCHEMA_REGISTRY_RESOURCE_ID
+            az iot ops upgrade --name myinstance -g myresourcegroup --ops-version x.y.z --ops-train preview
```
Comment (vilit1) [inline azext_edge/edge/_help.py:597]: do we expect external customers to use this?

File: azext_edge/edge/providers/orchestration/upgrade2.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    def analyze_cluster(self, **override_kwargs: dict) -> "ClusterUpgradeState":
+        with Progress(
+            SpinnerColumn("star"),
+            *Progress.get_default_columns(),
+            "Elapsed:",
+            TimeElapsedColumn(),
+            transient=True,
+            disable=bool(self.no_progress),
+        ) as progress:
+            _ = progress.add_task("Analyzing cluster...", total=None)
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/upgrade2.py:102]: can you just do `progress.add_task("Analyzing cluster...", total=None)` - the wildcard is extra

File: azext_edge/edge/_help.py (hunk 1 above)
Comment (digimaun) [reply azext_edge/edge/_help.py:597]: Yes

File: azext_edge/edge/util/common.py [hunk 3]
```diff
@@ -19,6 +19,20 @@
...
+    """
+    Parses kvp nargs into a dict handling values of null and empty string.
+    """
+    result = {}
+    if not kvp_nargs:
+        return result
+
+    for item in kvp_nargs:
+        key, sep, value = item.partition("=")
+        result[key] = value if sep else None
```
Comment (vilit1) [inline azext_edge/edge/util/common.py:32]: This accepts inputs like "potato" -> {"potato": None}

I thought the og command (below) did not take these cases for a reason? Is that reason still valid

(PS: I do like the simplicity of this more - just double checking)

Comment (digimaun) [reply azext_edge/edge/util/common.py:32]: Correct. This function would replace most usage of the prior function. The null case is useful and necessary, for example it is used to delete a PATCH configuration setting. It is better in every way. Look at the unit test for all the cases.

File: azext_edge/tests/edge/orchestration/test_upgrade2_unit.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    get_base_endpoint,
+    get_mock_resource,
+)
+from .resources.test_instances_unit import (
+    get_instance_endpoint,
+    get_mock_cl_record,
+    get_mock_instance_record,
+)
+
+T = TypeVar("T", bound="UpgradeScenario")
```
Comment (vilit1) [inline azext_edge/tests/edge/orchestration/test_upgrade2_unit.py:47]: prob use something more descriptive than just T (you can do US or UpScen or something if you want to keep it short)

File: azext_edge/edge/providers/orchestration/common.py [hunk 5]
```diff
@@ -40,6 +40,34 @@
...
+    "iotOperations": "ops",
+}
+
+EXTENSION_ALIAS_TO_TYPE_MAP = {
+    "plat": EXTENSION_TYPE_PLATFORM,
+    "osm": EXTENSION_TYPE_OSM,
+    "ssc": EXTENSION_TYPE_SSC,
+    "acs": EXTENSION_TYPE_ACS,
+    "ops": EXTENSION_TYPE_OPS,
+}
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/common.py:65]: I like how this becomes a connected single direction triangle

File: azext_edge/edge/commands_edge.py [hunk 6]
```diff
@@ -220,24 +220,52 @@ def create_instance(
...
 
-def upgrade(
+def upgrade_instance(
     cmd,
     resource_group_name: str,
     instance_name: str,
-    schema_registry_resource_id: Optional[str] = None,
     no_progress: Optional[bool] = None,
     confirm_yes: Optional[bool] = None,
+    ops_config: Optional[List[str]] = None,
```
Comment (vilit1) [inline azext_edge/edge/commands_edge.py:229]: do we have a list of key value pairs that are actually accepted by the service? or do we assume that the service teams are the main consumers for this (and they would know the kv pairs themselves)

File: azext_edge/edge/providers/orchestration/upgrade2.py [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+        self.no_progress = no_progress
+        self.instances = Instances(self.cmd)
+        self.resource_map = self.instances.get_resource_map(
+            self.instances.show(name=self.instance_name, resource_group_name=self.resource_group_name)
+        )
+        self.targets = InitTargets(
+            cluster_name=self.resource_map.connected_cluster.cluster_name, resource_group_name=resource_group_name
+        )
+        self.init_version_map: Dict[str, dict] = {}
+        self.init_version_map.update(self.targets.get_extension_versions())
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/upgrade2.py:90]: nit: could combine this line with the one above

File: azext_edge/edge/commands_edge.py (hunk 6 above)
Comment (digimaun) [reply azext_edge/edge/commands_edge.py:229]: Some user kvp's are described in AIO learn documentation intended for end users. Kvp's can be provided and used both internally and externally (power uses, troubleshooting etc) for whatever reasons.

File: azext_edge/edge/providers/orchestration/upgrade2.py [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+        return bool(self.override.version) or (
+            self.desired_version[0] and version.parse(self.desired_version[0]) > version.parse(self.current_version[0])
+        )
+
+    def _has_delta_in_train(self) -> bool:
+        return bool(self.override.train) or (
+            self.desired_version[1] and self.desired_version[1].lower() != self.current_version[1].lower()
+        )
+
+    def _has_delta_in_config(self) -> bool:
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/upgrade2.py:291]: are we going to check the desired config states?

File: azext_edge/edge/providers/orchestration/upgrade2.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    ):
+        self.cmd = cmd
+        self.instance_name = instance_name
+        self.resource_group_name = resource_group_name
+        self.no_progress = no_progress
+        self.instances = Instances(self.cmd)
+        self.resource_map = self.instances.get_resource_map(
+            self.instances.show(name=self.instance_name, resource_group_name=self.resource_group_name)
+        )
+        self.targets = InitTargets(
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/upgrade2.py:86]: nit: this is neat over importing the template and then crying

File: azext_edge/edge/providers/orchestration/upgrade2.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+        self.override_map = override_map
+        self.extension_upgrades = self.refresh_upgrade_state()
+
+    def has_upgrades(self) -> bool:
+        return any(ext_state.can_upgrade() for ext_state in self.extension_upgrades)
+
+    def refresh_upgrade_state(self) -> List["ExtensionUpgradeState"]:
+        ext_queue: List["ExtensionUpgradeState"] = []
+
+        # TODO @digimaun - deterine further pre-checks.
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/upgrade2.py:214]: are we handling system extensions or not yet (TODO then?)

Comment (digimaun) [reply azext_edge/edge/providers/orchestration/upgrade2.py:214]: Yea, need input for those decisions. Will include in description.

File: azext_edge/edge/providers/orchestration/resources/clusters.py [hunk 11]
```diff
@@ -58,6 +58,7 @@ def update_cluster_extension(
         cluster_name: str,
         extension_name: str,
         update_payload: dict,
+        **client_kwargs,
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/resources/clusters.py:61]: nit: can prob remove line 54

File: azext_edge/edge/providers/orchestration/upgrade2.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
+# coding=utf-8
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/upgrade2.py:1]: if you delete the og file - may as well just rename this to upgrade.py

File: azext_edge/tests/edge/orchestration/test_upgrade2_unit.py [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+
+
+def get_cluster_extensions_endpoint(resource_group_name: str, cluster_name: str = "mycluster") -> dict:
+    resource_path = f"/connectedClusters/{cluster_name}/providers/Microsoft.KubernetesConfiguration/extensions"
+    return get_base_endpoint(
+        resource_group_name=resource_group_name,
+        resource_path=resource_path,
+        resource_provider="Microsoft.Kubernetes",
+        api_version=CLUSTER_EXTENSIONS_API_VERSION,
+    )
```
Comment (vilit1) [inline azext_edge/tests/edge/orchestration/test_upgrade2_unit.py:84]: resource group casing change only for base cluster?

Comment (digimaun) [reply azext_edge/tests/edge/orchestration/test_upgrade2_unit.py:84]: It is due to the rest api spec, which leads into the generated client, something I don't want to alter.

File: azext_edge/edge/providers/orchestration/upgrade2.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+    def _has_delta_in_train(self) -> bool:
+        return bool(self.override.train) or (
+            self.desired_version[1] and self.desired_version[1].lower() != self.current_version[1].lower()
+        )
+
+    def _has_delta_in_config(self) -> bool:
+        return bool(self.override.config)
+
+
+def get_default_table() -> Table:
```
Comment (vilit1) [inline azext_edge/edge/providers/orchestration/upgrade2.py:295]: nit: please add screenshot of examples

File: azext_edge/edge/commands_edge.py (hunk 6 above)
Comment (vilit1) [reply azext_edge/edge/commands_edge.py:229]: would be nice to include these in the long help but not necessary

File: azext_edge/edge/providers/orchestration/upgrade2.py (hunk 12 above)
Comment (vilit1) [reply azext_edge/edge/providers/orchestration/upgrade2.py:1]: if you are doing this for the git compare - you can do it as a last push before merge

Comment (digimaun) [reply azext_edge/edge/providers/orchestration/upgrade2.py:1]: Yes this is to simplify the review.

File: azext_edge/edge/providers/orchestration/upgrade2.py (hunk 8 above)
Comment (digimaun) [reply azext_edge/edge/providers/orchestration/upgrade2.py:291]: You are talking about some delta between extension X 1.0 -> X 2.0 right? Lets cross that bridge when we get there.

**Review by c-ryan-k — APPROVED** [review]

File: azext_edge/edge/providers/orchestration/upgrade2.py (hunk 2 above)
Comment (c-ryan-k) [reply azext_edge/edge/providers/orchestration/upgrade2.py:102]: I can't remember if it was, or still is, a thing - but I think there was a linter rule regarding calling functions that return a value without assigning that value to something?

File: azext_edge/edge/providers/orchestration/upgrade2.py (hunk 12 above)
Comment (digimaun) [reply azext_edge/edge/providers/orchestration/upgrade2.py:1]: Going to do in subsequent PR.

**Pull request merged**